src/core/net/srp_server.cpp

```diff
@@ -0,0 +1,1577 @@
+        {
+            otLogInfoSrp("server: LEASE of host %s expires", host->GetFullName());
+
+            // If the host expires, delete all resources of this host and its services.
+            host->ClearResources();
+            for (auto service = host->GetServices(); service != nullptr; service = service->GetNext())
+            {
+                service->SetDeleted(true);
+                service->ClearResources();
+            }
+
+            earliestExpireTime = OT_MIN(earliestExpireTime, host->GetKeyExpireTime());
+        }
+        else
+        {
+            // The host doesn't expire, check if any service expires or is explicitly removed.
+
+            OT_ASSERT(!host->IsDeleted());
+
+            auto service = host->GetServices();
+
+            earliestExpireTime = OT_MIN(earliestExpireTime, host->GetExpireTime());
+
+            while (service != nullptr)
+            {
+                auto nextService = service->GetNext();
+
+                if (service->IsDeleted())
+                {
+                    // The service has been deleted but the name retains.
+                    earliestExpireTime = OT_MIN(earliestExpireTime, service->GetKeyExpireTime());
+                }
+                else if (service->GetExpireTime() <= now)
+                {
+                    otLogInfoSrp("server: LEASE of service %s expires", service->GetFullName());
+
+                    // The service gets expired, delete it.
+
+                    service->SetDeleted(true);
+                    service->ClearResources();
+
+                    earliestExpireTime = OT_MIN(earliestExpireTime, service->GetKeyExpireTime());
+                }
+                else
+                {
+                    earliestExpireTime = OT_MIN(earliestExpireTime, service->GetExpireTime());
+                }
+
+                service = nextService;
+            }
+        }
+
+        host = nextHost;
+    }
+
+    if (earliestExpireTime != now.GetDistantFuture())
+    {
+        if (!mLeaseTimer.IsRunning() || earliestExpireTime <= mLeaseTimer.GetFireTime())
+        {
+            otLogInfoSrp("server: lease timer is scheduled in %u seconds", (earliestExpireTime - now) / 1000);
+            mLeaseTimer.StartAt(earliestExpireTime, 0);
+        }
+    }
+    else
+    {
+        otLogInfoSrp("server: lease timer is stopped");
+        mLeaseTimer.Stop();
+    }
+}
+
+void Server::HandleOutstandingUpdatesTimer(Timer &aTimer)
+{
+    aTimer.GetOwner<Server>().HandleOutstandingUpdatesTimer();
+}
+
+void Server::HandleOutstandingUpdatesTimer(void)
+{
+    while (!mOutstandingUpdates.IsEmpty() && mOutstandingUpdates.GetTail()->GetExpireTime() <= TimerMilli::GetNow())
+    {
+        HandleAdvertisingResult(mOutstandingUpdates.GetTail(), OT_ERROR_RESPONSE_TIMEOUT);
+    }
+}
+
+Server::Service *Server::Service::New(const char *aFullName)
+{
+    void *   buf;
+    Service *service = nullptr;
+
+    buf = Instance::Get().HeapCAlloc(1, sizeof(Service));
```

Comment:
The `Instance::Get()` is only available in single instance case.

FWIW, we still support/have the multi-instance case within OT, so would be good to have SRP server support it as well.
Here is my suggestion:
- You can pass `instance` as paramter to `New()`
- Make the `Servcie` and `Host` be `InstanceLocator`. 
- Pass the `instance` in `new (buf) Service(instance);`

Reply:
SGTM, updated.